File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/spiders/SpiderMind.java
package com.nyrds.pixeldungeon.mobs.spiders;

import com.nyrds.pixeldungeon.ai.Hunting;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.buffs.Barkskin;
import com.watabou.pixeldungeon.actors.buffs.Blessed;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Fury;
import com.watabou.pixeldungeon.actors.buffs.Speed;
import com.watabou.pixeldungeon.actors.mobs.Mob;
import com.watabou.pixeldungeon.effects.CellEmitter;
import com.watabou.pixeldungeon.effects.particles.ShaftParticle;
import com.watabou.pixeldungeon.items.food.MysteryMeat;
import com.watabou.pixeldungeon.plants.Earthroot;
import com.watabou.pixeldungeon.plants.Sungrass;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

import lombok.val;

public class SpiderMind extends Mob {

	private static final String[] BuffsForFriends = {
		Speed.class.getSimpleName(),
		Barkskin.class.getSimpleName(),
		Blessed.class.getSimpleName(),
		Sungrass.Health.class.getSimpleName(),
		Earthroot.Armor.class.getSimpleName(),
		"ManaShield",
		Fury.class.getSimpleName()
	};

	public SpiderMind() {
		hp(ht(5));
		baseDefenseSkill = 1;
		baseAttackSkill  = 10;
		baseSpeed = 1f;
		dmgMin = 0;
		dmgMax = 0;
		dr = 0;
		
		expForKill = 6;
		maxLvl = 9;
		
		loot(new MysteryMeat(), 0.067f);
	}
	
	@Override
    public boolean canAttack(@NotNull Char enemy) {
		return CharUtils.canDoOnlyRangedAttack(this, enemy);
	}
	
	@Override
	public int zapProc(@NotNull Char enemy, int damage ) {

		for (val mob : level().mobs){
			if(mob != this) {
				int mobPos = mob.getPos();
				if(level().fieldOfView[mobPos] && mob.friendly(this)) {
					Buff.prolong(mob, Random.oneOf(BuffsForFriends), 3);

					if (Dungeon.isCellVisible(mobPos)) {
						CellEmitter.get(mobPos).start(ShaftParticle.FACTORY, 0.2f, 3);
					}

					break;
				}
			}
		}

		return 0;
	}
	
	@Override
	public boolean getCloser(int target) {
		if (getState() instanceof Hunting) {
			return enemySeen && getFurther( target );
		} else {
			return super.getCloser( target );
		}
	}
}

File: scripts/mobs/SpiderMind.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

-- java SpiderMind (batch 10): harmless buff-bot - its zap buffs the first
-- friendly mob in sight (3 turns) instead of hurting the enemy, and it
-- retreats while hunting (aiState Kite + kiteNeverApproach) keeping its
-- ranged line.
return mob.init{
    stats = function(self)
        mob.restoreData(self).kiteNeverApproach = true
    end,

    zapProc = function(self, enemy, dmg)
        local level = RPD.Dungeon.level
        local buffs = {"Speed", "Barkskin", "Blessed", "Health", "Armor", "ManaShield", "Fury"}
        local mobs = level:getCopyOfMobsArray()

        for i = 1, #mobs do
            local chr = mobs[i]
            if chr ~= self then
                local chrPos = chr:getPos()
                -- luaj arrays are 1-based: fieldOfView[pos + 1] == java [pos]
                if level.fieldOfView[chrPos + 1] and chr:friendly(self) then
                    RPD.Buffs.Buff:prolong(chr, buffs[math.random(#buffs)], 3)
                    if RPD.Dungeon:isCellVisible(chrPos) then
                        RPD.Sfx.CellEmitter:get(chrPos):start(RPD.Sfx.ShaftParticle.FACTORY, 0.2, 3)
                    end
                    break
                end
            end
        end
        return 0
    end,

}
